### PR TITLE
implement new IO interfaces

### PR DESCRIPTION
**Pull request by bicycle1885**
This implements new IO interfaces discussed in #288. While working on it, I found a few suspicious problems on FASTQ and BigBed, but those are beyond the scope of this pull request. I'll make separated pull requests to fix them.

### CODE REVIEW

File: src/align/hts/bam/reader.jl [hunk 1]
```diff
@@ -10,6 +10,19 @@ type BAMReader <: Bio.IO.AbstractReader
     index::Nullable{BAI}
 end
 
+function BAMReader(input::IO; index=nothing)
+    if isa(index, AbstractString)
```
Comment (kdmurray91) [inline src/align/hts/bam/reader.jl]: String or AbstractString?

File: src/intervals/deprecated.jl [hunk 2]
```diff
@@ -0,0 +1,11 @@
+immutable BED <: Bio.IO.FileFormat end
+
+export BED
+
+function Base.open{F<:Union{BED,BigBed}}(::AbstractString, ::Type{F})
```
Comment (kdmurray91) [inline src/intervals/deprecated.jl]: Is it not worth defining these for all abstract `FileFormat`s under Bio.IO?

File: test/seq/runtests.jl [hunk 3]
```diff
@@ -3092,7 +3100,14 @@ end
                contains(tags, "comments") || contains(tags, "ambiguity")
                 continue
             end
-            test_fastq_parse(joinpath(path, specimen["filename"]), valid)
+            filename = specimen["filename"]
+            # FIXME
```
Comment (kdmurray91) [inline test/seq/runtests.jl]: Should this get fixed in this PR?

**Comment by kdmurray91** [comment]
@bicycle1885 LGTM. Couple of comments made.

File: src/align/hts/bam/reader.jl (hunk 1 above)
Comment (bicycle1885) [inline src/align/hts/bam/reader.jl]: `String` is a subtype of `AbstractString` so `isa(index, AbstractString)` returns `true` when `index` is `String`.

File: src/intervals/deprecated.jl (hunk 2 above)
Comment (bicycle1885) [inline src/intervals/deprecated.jl]: Yes! That sounds more elegant.

File: test/seq/runtests.jl (hunk 3 above)
Comment (bicycle1885) [inline test/seq/runtests.jl]: No, I'm going to fix it shortly in the next pull request.

**Comment by bicycle1885** [comment]
I will merge this soon if no objections.

**Pull request merged**